hazelcast/src/hazelcast/client/network.cpp

```diff
@@ -440,51 +455,51 @@ namespace hazelcast {
             }
 
             bool ClientConnectionManagerImpl::do_connect_to_cluster() {
-                int attempt = 0;
-                std::unordered_set<address> triedAddresses;
-
-                while (attempt < connection_attempt_limit_) {
-                    attempt++;
-                    auto nextTryTime = std::chrono::steady_clock::now() + connection_attempt_period_;
+                std::unordered_set<address> tried_addresses;
+                wait_strategy_.reset();
+
+                do {
+                    std::unordered_set<address> tried_addresses_per_attempt;
+                    auto member_list = client_.get_client_cluster_service().get_member_list();
+                    if (shuffle_member_list_) {
+                        shuffle(member_list);
+                    }
 
-                    for (const address &address : get_possible_member_addresses()) {
+                    //try to connect to a member in the member list first
+                    for (const auto &m : member_list) {
                         check_client_active();
-                        triedAddresses.insert(address);
-                        auto connection = connect(address);
+                        tried_addresses_per_attempt.insert(m.get_address());
+                        auto connection = connect<member>(m, [=](const member &server) {
+                            return get_or_connect_to_member(server);
+                        });
                         if (connection) {
                             return true;
                         }
                     }
+                    //try to connect to a member given via config(explicit config/discovery mechanisms)
+                    for (const address &server_address : get_possible_member_addresses()) {
+                        check_client_active();
+                        if (!tried_addresses_per_attempt.insert(server_address).second) {
+                            //if we can not add it means that it is already tried to be connected with the member list
+                            continue;
+                        }
 
-                    // If the address providers load no addresses (which seems to be possible), then the above loop is not entered
+                        auto connection = connect<address>(server_address, [=](const address &target) {
+                            return get_or_connect_to_address(target);
+                        });
+                        if (connection) {
+                            return true;
+                        }
+                    }
```

Comment:
The method `ClientConectionManagerImpl::connect` doesn't actually connect anywhere. It just wraps the given function call with a try-catch and prints a log statement if the function throws an exception. I believe it's just there to avoid duplicating the same logic for these two for-loops. But still, these two for loops share the same piece of code.

I think I have a better solution. First, make `get_or_connect_to_address` and `get_or_connect_to_member` specialization of the same function template, like `get_or_connect<address>` and `get_or_connect<member>`. Then, write another method for the logic in the for loop:

```C++
template<typename T>
shared_ptr<Connection> try_connect(T target) {
   check_client_active()
   if(!tried_addresses_per_attempt.insert(server_address).second)
     return false;
   // same as in connect<>:
   try {
      return get_or_connect<T>(target)
   } except (...) {
      HZ_LOG(...)
   }
}

```

Now we can convert the for loops into:

```C++
for (auto m: member_list) {
   auto conn = try_connect(m);
   if (conn) return true;
}

for (auto a: addresses) {
   auto conn = try_connect(a);
   if (conn) return true;
}
```

Reply:
Changed.